### ISSUE TITLE
Type 'ArrayBufferView' is not assignable to type 'DataView' 

### ISSUE THREAD

**Issue by mhart** [microsoft/TypeScript#38715]
**TypeScript Version:**  Nightly

**Search Terms:**

ArrayBufferView

**Expected behavior:**

According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/ArrayBufferView), ArrayBufferView is one of these types:

- Int8Array,
- Uint8Array,
- Uint8ClampedArray,
- Int16Array,
- Uint16Array,
- Int32Array,
- Uint32Array,
- Float32Array,
- Float64Array or
- DataView.

However, in lib.es5.d.ts it's just an interface.

This would be fine if it were used consistently, but it appears that it's used in some places (eg `BodyInit`) but then in others the actual types from above are listed (eg `crypto.subtle.digest`). This means, for example, that going from `BodyInit` to `crypto.subtle.digest` (while filtering out incompatible types) is impossible.

According to these links:

1. https://developer.mozilla.org/en-US/docs/Web/API/Request/Request#Syntax
2. https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#Syntax

This code should compile:

```ts
function bodyHash(body: BodyInit) {
    if (typeof body === 'string' ||
        body instanceof Blob ||
        body instanceof FormData ||
        body instanceof URLSearchParams ||
        body instanceof ReadableStream) {
            throw new Error('Only ArrayBuffer or ArrayBufferView allowed')
    }
    // error 2345: Type 'ArrayBufferView' is not assignable to type 'DataView'
    let result = crypto.subtle.digest('SHA-256', body)
}
```

**Actual behavior:**

Code fails to compile with type errors

**Related Issues:**

Possibly https://github.com/microsoft/TypeScript/issues/15402

**Code**
```ts
// https://developer.mozilla.org/en-US/docs/Web/API/ArrayBufferView
type ActualArrayBufferView = Int8Array | Uint8Array | Uint8ClampedArray | Int16Array | Uint16Array | Int32Array | Uint32Array | Float32Array | Float64Array | DataView

function actual(arr: ActualArrayBufferView) {
    // https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#Syntax
    // works as it should with no error
    let result = crypto.subtle.digest('SHA-256', arr)
}

function builtin(arr: ArrayBufferView) {
    // error 2345: Type 'ArrayBufferView' is missing the following properties from type 'DataView': getFloat32, getFloat64, getInt8, getInt16, and 13 more
    actual(arr)

    // error 2345: Type 'ArrayBufferView' is not assignable to type 'DataView'
    let result = crypto.subtle.digest('SHA-256', arr)
}

// Real-world example
function bodyHash(body: BodyInit) {
    if (typeof body === 'string' ||
        body instanceof Blob ||
        body instanceof FormData ||
        body instanceof URLSearchParams ||
        body instanceof ReadableStream) {
            throw new Error('Only ArrayBuffer or ArrayBufferView allowed')
    }
    // error 2345: Type 'ArrayBufferView' is not assignable to type 'DataView'
    let result = crypto.subtle.digest('SHA-256', body)
}
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
function actual(arr) {
    // https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#Syntax
    // works as it should with no error
    let result = crypto.subtle.digest('SHA-256', arr);
}
function builtin(arr) {
    // error 2345: Type 'ArrayBufferView' is missing the following properties from type 'DataView': getFloat32, getFloat64, getInt8, getInt16, and 13 more
    actual(arr);
    // error 2345: Type 'ArrayBufferView' is not assignable to type 'DataView'
    let result = crypto.subtle.digest('SHA-256', arr);
}
// Real-world example
function bodyHash(body) {
    if (typeof body === 'string' ||
        body instanceof Blob ||
        body instanceof FormData ||
        body instanceof URLSearchParams ||
        body instanceof ReadableStream) {
        throw new Error('Only ArrayBuffer or ArrayBufferView allowed');
    }
    // error 2345: Type 'ArrayBufferView' is not assignable to type 'DataView'
    let result = crypto.subtle.digest('SHA-256', body);
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?ts=4.0.0-dev.20200520#code/PTAEAsBdIBwZwFwgCYFMBuqA2B7GqAnAOgFscAvASyywEMicCBzYVAOwFoBVAZWGRwBjOMADqqAEbAAggAUAkjIIFaATwBCAVwBm2wgDVKqAO4AoSKvyhpgyJtpZpytVt0Gjx0AF5Q8tpAAOJxVVUAAfUC5KfyDnUIiomIBhOhJ8ZGC1cN9-AEYANkz4yOjIAqLsv0gAZgAmCoTSuobQADFcWhr6uOz2nE78gBYWgBFO2kMTU1NtTTZbShw2UFpbeywAClplBGs1hyLXPQJJ4wBKUABvU1Bb0BAIaHgkfgxsPEJSCmo6BmZWTi8fhCETiKRyRQ8TQSSBYVBJAiWSA4fiUJioOCQADEPFU-loAA8bncHsZGABrOArKmUSCgODgHCaLDIUDGWngUBsHCgQgERjE25wukEDHMuk+QSImDIohwaGw1BEZBojGQDYAch4AAlpBxagBWfIagA0K2UZ1MAF9prN5pBFssJJpqA62FsdtY4kd3CYLtc7vcwHzGKBatVBgbdgAVSyoUAaw46Y6nDWgShUkgZuDRJigSDgePaHA0HDsth5mD8-AEB0Y0DafkkfNxhNjSATDwa3boyB9Tp1M29-uQIZD1CQKoBceTvL5M20Nis3LVUBkUWClb7TbbAiWzcPEMEMMRqOgWNWRPe5O+4xpjNcnB02hwHNMNi0CRw-M8iyX9udiYGqbsKoCivKWASqAUpIjgcoKnCyqqpimo6nqhrGguFrWtMDwAEqoA4HBkgQLK8gStBpHCMxzAsSygBIODIKo2ovuAGyMcxuzqExqh+LS-qbpQ2igBsf6oDgImcaEXiyQmmIELmaZhGEm6BtJ6ZsJii6CBJInqLgEjhKpgbqbxmnafMeltIwJAAcZal3Bp0SWbpkmRHhAAyPCEQQgjgLI2yUVSKmObczlaR2VnuQRtDIJ+cI8JAoqUYJpnpfm4D8p4bAmKAACiyiMJqADybBYKESZuMeoZVSmHgrKWxioMgGqWoGNqBoeRXHuGkYxq2V4hD6JxdumVLcs+r5oh+X7xsiLb-uMqYgROYFipB3jQdKsryjCiEquiKFarq+pGqaDG8ZaVpAA)

**Comment by mazerab** [comment]
Hello, I am running into this exact issue and was wondering if there is any workaround one could apply. Thank you, Bastien

**Comment by jakebailey** [comment]
This was fixed in #44684, and is in TS 4.4+.
